File: sdk/golang/sfa/cli_test.go
package sfa

import (
	"testing"
)

func TestParseArgsStandardFlags(t *testing.T) {
	args, err := parseArgs([]string{
		"--help",
	}, nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !args.Flags.Help {
		t.Error("expected --help to be true")
	}
}

func TestParseArgsVersion(t *testing.T) {
	args, err := parseArgs([]string{"--version"}, nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !args.Flags.Version {
		t.Error("expected --version to be true")
	}
}

func TestParseArgsDescribe(t *testing.T) {
	args, err := parseArgs([]string{"--describe"}, nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !args.Flags.Describe {
		t.Error("expected --describe to be true")
	}
}

func TestParseArgsAllStandardFlags(t *testing.T) {
	args, err := parseArgs([]string{
		"--verbose", "--quiet", "--output-format", "json",
		"--timeout", "60", "--setup", "--no-log",
		"--max-depth", "3", "--services-down", "--yes",
		"--non-interactive", "--mcp",
		"--context", "hello world",
		"--context-file", "/tmp/ctx.txt",
	}, nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if !args.Flags.Verbose {
		t.Error("expected verbose")
	}
	if !args.Flags.Quiet {
		t.Error("expected quiet")
	}
	if args.Flags.OutputFormat != OutputJSON {
		t.Errorf("expected json, got %s", args.Flags.OutputFormat)
	}
	if args.Flags.Timeout != 60 {
		t.Errorf("expected timeout 60, got %d", args.Flags.Timeout)
	}
	if !args.Flags.Setup {
		t.Error("expected setup")
	}
	if !args.Flags.NoLog {
		t.Error("expected no-log")
	}
	if args.Flags.MaxDepth != 3 {
		t.Errorf("expected max-depth 3, got %d", args.Flags.MaxDepth)
	}
	if !args.Flags.ServicesDown {
		t.Error("expected services-down")
	}
	if !args.Flags.Yes {
		t.Error("expected yes")
	}
	if !args.Flags.NonInteractive {
		t.Error("expected non-interactive")
	}
	if !args.Flags.MCP {
		t.Error("expected mcp")
	}
	if args.Flags.Context != "hello world" {
		t.Errorf("expected context 'hello world', got %q", args.Flags.Context)
	}
	if args.Flags.ContextFile != "/tmp/ctx.txt" {
		t.Errorf("expected context-file /tmp/ctx.txt, got %q", args.Flags.ContextFile)
	}
}

func TestParseArgsDefaults(t *testing.T) {
	args, err := parseArgs([]string{}, nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if args.Flags.Timeout != 120 {
		t.Errorf("expected default timeout 120, got %d", args.Flags.Timeout)
	}
	if args.Flags.MaxDepth != 5 {
		t.Errorf("expected default max-depth 5, got %d", args.Flags.MaxDepth)
	}
	if args.Flags.OutputFormat != OutputText {
		t.Errorf("expected default output-format text, got %s", args.Flags.OutputFormat)
	}
}

func TestParseArgsCustomStringOption(t *testing.T) {
	opts := []OptionDef{
		{Name: "model", Type: "string", Default: "gpt-4", Description: "Model to use"},
	}
	args, err := parseArgs([]string{"--model", "claude-3"}, opts)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if args.Custom["model"] != "claude-3" {
		t.Errorf("expected model claude-3, got %v", args.Custom["model"])
	}
}

func TestParseArgsCustomBoolOption(t *testing.T) {
	opts := []OptionDef{
		{Name: "dry-run", Type: "boolean", Description: "Dry run mode"},
	}
	args, err := parseArgs([]string{"--dry-run"}, opts)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if args.Custom["dry-run"] != true {
		t.Errorf("expected dry-run true, got %v", args.Custom["dry-run"])
	}
}

func TestParseArgsCustomNumberOption(t *testing.T) {
	opts := []OptionDef{
		{Name: "count", Type: "number", Default: 10, Description: "Count"},
	}
	args, err := parseArgs([]string{"--count", "42"}, opts)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if args.Custom["count"] != 42 {
		t.Errorf("expected count 42, got %v", args.Custom["count"])
	}
}

func TestParseArgsCustomOptionWithAlias(t *testing.T) {
	opts := []OptionDef{
		{Name: "model", Alias: "m", Type: "string", Description: "Model"},
	}
	args, err := parseArgs([]string{"-m", "claude"}, opts)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if args.Custom["model"] != "claude" {
		t.Errorf("expected model claude, got %v", args.Custom["model"])
	}
}

func TestReadInputFromContext(t *testing.T) {
	input, err := readInput(StandardFlags{Context: "test data"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if input != "test data" {
		t.Errorf("expected 'test data', got %q", input)
	}
}

func TestReadInputFromContextFile(t *testing.T) {
	tmpDir := t.TempDir()
	f := tmpDir + "/input.txt"
	if err := writeTestFile(f, "file content"); err != nil {
		t.Fatal(err)
	}

	input, err := readInput(StandardFlags{ContextFile: f})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if input != "file content" {
		t.Errorf("expected 'file content', got %q", input)
	}
}

func TestReadInputNoInput(t *testing.T) {
	input, err := readInput(StandardFlags{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if input != "" {
		t.Errorf("expected empty input, got %q", input)
	}
}

func TestGenerateHelp(t *testing.T) {
	def := &AgentDef{
		Name:        "test-agent",
		Version:     "1.0.0",
		Description: "A test agent",
		Options: []OptionDef{
			{Name: "model", Alias: "m", Description: "Model to use"},
		},
		Env: []EnvDef{
			{Name: "API_KEY", Required: true, Description: "API key"},
		},
		Examples: []string{
			"echo 'hello' | test-agent",
		},
	}

	help := generateHelp(def)
	if help == "" {
		t.Error("expected non-empty help text")
	}

	// Check key sections exist
	for _, want := range []string{"test-agent v1.0.0", "USAGE:", "OPTIONS:", "AGENT OPTIONS:", "--model", "ENVIRONMENT VARIABLES:", "API_KEY", "EXAMPLES:"} {
		if !contains(help, want) {
			t.Errorf("help missing %q", want)
		}
	}
}

func TestGenerateDescribe(t *testing.T) {
	def := &AgentDef{
		Name:        "test-agent",
		Version:     "1.0.0",
		Description: "A test agent",
		TrustLevel:  TrustNetwork,
		Env: []EnvDef{
			{Name: "API_KEY", Required: true, Secret: true, Description: "API key"},
		},
	}
	env := map[string]string{"API_KEY": "secret123"}
	secrets := map[string]bool{"API_KEY": true}

	desc := generateDescribe(def, env, secrets)

	if desc["name"] != "test-agent" {
		t.Errorf("expected name test-agent, got %v", desc["name"])
	}
	if desc["trustLevel"] != "network" {
		t.Errorf("expected trustLevel network, got %v", desc["trustLevel"])
	}

	// Check secret masking in describe
	envList := desc["env"].([]map[string]any)
	if envList[0]["value"] != "***" {
		t.Errorf("expected secret value masked, got %v", envList[0]["value"])
	}
}

func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(s) > 0 && containsSubstr(s, substr))
}

func containsSubstr(s, sub string) bool {
	for i := 0; i <= len(s)-len(sub); i++ {
		if s[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}

File: tests/sdk/cli.test.ts
import { test, expect, describe } from "bun:test";
import { parseArgs } from "../../@sfa/sdk/cli";

describe("parseArgs", () => {
  describe("standard flags", () => {
    test("parses --help as boolean", () => {
      const result = parseArgs(["--help"]);
      expect(result.flags.help).toBe(true);
    });

    test("parses --version as boolean", () => {
      const result = parseArgs(["--version"]);
      expect(result.flags.version).toBe(true);
    });

    test("parses --verbose as boolean", () => {
      const result = parseArgs(["--verbose"]);
      expect(result.flags.verbose).toBe(true);
    });

    test("parses --quiet as boolean", () => {
      const result = parseArgs(["--quiet"]);
      expect(result.flags.quiet).toBe(true);
    });

    test("parses --describe as boolean", () => {
      const result = parseArgs(["--describe"]);
      expect(result.flags.describe).toBe(true);
    });

    test("parses --setup as boolean", () => {
      const result = parseArgs(["--setup"]);
      expect(result.flags.setup).toBe(true);
    });

    test("parses --no-log as boolean", () => {
      const result = parseArgs(["--no-log"]);
      expect(result.flags["no-log"]).toBe(true);
    });

    test("parses --services-down as boolean", () => {
      const result = parseArgs(["--services-down"]);
      expect(result.flags["services-down"]).toBe(true);
    });

    test("parses --yes as boolean", () => {
      const result = parseArgs(["--yes"]);
      expect(result.flags.yes).toBe(true);
    });

    test("parses --non-interactive as boolean", () => {
      const result = parseArgs(["--non-interactive"]);
      expect(result.flags["non-interactive"]).toBe(true);
    });

    test("parses --mcp as boolean", () => {
      const result = parseArgs(["--mcp"]);
      expect(result.flags.mcp).toBe(true);
    });

    test("parses --output-format with space-separated value", () => {
      const result = parseArgs(["--output-format", "json"]);
      expect(result.flags["output-format"]).toBe("json");
    });

    test("parses --output-format with = value", () => {
      const result = parseArgs(["--output-format=json"]);
      expect(result.flags["output-format"]).toBe("json");
    });

    test("parses --timeout as number", () => {
      const result = parseArgs(["--timeout", "60"]);
      expect(result.flags.timeout).toBe(60);
    });

    test("parses --timeout with = value", () => {
      const result = parseArgs(["--timeout=30"]);
      expect(result.flags.timeout).toBe(30);
    });

    test("parses --max-depth as number", () => {
      const result = parseArgs(["--max-depth", "3"]);
      expect(result.flags["max-depth"]).toBe(3);
    });

    test("parses --context as string", () => {
      const result = parseArgs(["--context", "some input"]);
      expect(result.flags.context).toBe("some input");
    });

    test("parses --context-file as string", () => {
      const result = parseArgs(["--context-file", "/tmp/input.txt"]);
      expect(result.flags["context-file"]).toBe("/tmp/input.txt");
    });
  });

  describe("defaults", () => {
    test("boolean flags default to false", () => {
      const result = parseArgs([]);
      expect(result.flags.help).toBe(false);
      expect(result.flags.version).toBe(false);
      expect(result.flags.verbose).toBe(false);
      expect(result.flags.quiet).toBe(false);
      expect(result.flags.describe).toBe(false);
      expect(result.flags.setup).toBe(false);
      expect(result.flags["no-log"]).toBe(false);
      expect(result.flags["services-down"]).toBe(false);
      expect(result.flags.yes).toBe(false);
      expect(result.flags["non-interactive"]).toBe(false);
      expect(result.flags.mcp).toBe(false);
    });

    test("output-format defaults to text", () => {
      const result = parseArgs([]);
      expect(result.flags["output-format"]).toBe("text");
    });

    test("timeout defaults to 120", () => {
      const result = parseArgs([]);
      expect(result.flags.timeout).toBe(120);
    });

    test("max-depth defaults to 5", () => {
      const result = parseArgs([]);
      expect(result.flags["max-depth"]).toBe(5);
    });

    test("context and context-file default to undefined", () => {
      const result = parseArgs([]);
      expect(result.flags.context).toBeUndefined();
      expect(result.flags["context-file"]).toBeUndefined();
    });
  });

  describe("custom options", () => {
    test("parses custom string option", () => {
      const result = parseArgs(["--model", "gpt-4"], [
        { name: "model", description: "Model to use", type: "string" },
      ]);
      expect(result.custom.model).toBe("gpt-4");
    });

    test("parses custom number option", () => {
      const result = parseArgs(["--max-files", "10"], [
        { name: "max-files", description: "Max files", type: "number" },
      ]);
      expect(result.custom["max-files"]).toBe(10);
    });

    test("parses custom boolean option", () => {
      const result = parseArgs(["--dry-run"], [
        { name: "dry-run", description: "Dry run", type: "boolean" },
      ]);
      expect(result.custom["dry-run"]).toBe(true);
    });

    test("parses custom option with alias", () => {
      const result = parseArgs(["-m", "gpt-4"], [
        { name: "model", alias: "m", description: "Model to use", type: "string" },
      ]);
      expect(result.custom.model).toBe("gpt-4");
    });

    test("parses custom boolean option with alias", () => {
      const result = parseArgs(["-d"], [
        { name: "dry-run", alias: "d", description: "Dry run", type: "boolean" },
      ]);
      expect(result.custom["dry-run"]).toBe(true);
    });

    test("applies custom option defaults", () => {
      const result = parseArgs([], [
        { name: "model", description: "Model to use", type: "string", default: "gpt-3.5" },
      ]);
      expect(result.custom.model).toBe("gpt-3.5");
    });

    test("custom boolean option defaults to false", () => {
      const result = parseArgs([], [
        { name: "dry-run", description: "Dry run", type: "boolean" },
      ]);
      expect(result.custom["dry-run"]).toBe(false);
    });

    test("custom option with = syntax", () => {
      const result = parseArgs(["--model=claude"], [
        { name: "model", description: "Model to use", type: "string" },
      ]);
      expect(result.custom.model).toBe("claude");
    });
  });

  describe("positional arguments", () => {
    test("collects positional arguments", () => {
      const result = parseArgs(["file1.txt", "file2.txt"]);
      expect(result.positional).toEqual(["file1.txt", "file2.txt"]);
    });

    test("separates flags from positionals", () => {
      const result = parseArgs(["--verbose", "file.txt"]);
      expect(result.flags.verbose).toBe(true);
      expect(result.positional).toEqual(["file.txt"]);
    });

    test("-- stops flag parsing", () => {
      const result = parseArgs(["--verbose", "--", "--not-a-flag", "file.txt"]);
      expect(result.flags.verbose).toBe(true);
      expect(result.positional).toEqual(["--not-a-flag", "file.txt"]);
    });
  });

  describe("unknown flags", () => {
    test("collects unknown flags", () => {
      const result = parseArgs(["--unknown-flag"]);
      expect(result.unknown).toEqual(["--unknown-flag"]);
    });

    test("collects unknown short flags", () => {
      const result = parseArgs(["-x"]);
      expect(result.unknown).toEqual(["-x"]);
    });

    test("collects flag missing value as unknown", () => {
      const result = parseArgs(["--timeout"]);
      expect(result.unknown).toEqual(["--timeout"]);
    });
  });

  describe("edge cases", () => {
    test("empty argv returns defaults", () => {
      const result = parseArgs([]);
      expect(result.positional).toEqual([]);
      expect(result.unknown).toEqual([]);
    });

    test("multiple flags combined", () => {
      const result = parseArgs(["--verbose", "--quiet", "--timeout", "30", "--output-format", "json"]);
      expect(result.flags.verbose).toBe(true);
      expect(result.flags.quiet).toBe(true);
      expect(result.flags.timeout).toBe(30);
      expect(result.flags["output-format"]).toBe("json");
    });

    test("mixes standard and custom flags", () => {
      const result = parseArgs(["--verbose", "--model", "gpt-4", "input.txt"], [
        { name: "model", description: "Model", type: "string" },
      ]);
      expect(result.flags.verbose).toBe(true);
      expect(result.custom.model).toBe("gpt-4");
      expect(result.positional).toEqual(["input.txt"]);
    });
  });
});
